### ISSUE TITLE
Problem when modifying a Document pk

### ISSUE THREAD

**Issue by lafrech** [Scille/umongo#83]
If I create and commit a Document, then change its id, I get an error when committing:

```python
    def test_create(self, classroom_model):
        Student = classroom_model.Student
        john = Student(name='John Doe', birthday=datetime(1995, 12, 12))
        ret = john.commit()

        print(john.id)
        # 58580041f667370f03a677e1
        print(john.is_created)
        # True
        john.id = ObjectId("5672d47b1d41c88dcd37ef05")
        # No error
        print(john.is_created)
        # True
        john.commit()
        # umongo.exceptions.UpdateError: <pymongo.results.UpdateResult object at 0x7f19b14d4a20>
```

This is because in `commit()`, we fall in the _update_ case and pymongo searches for a doc with the new id.

umongo should either prevent changing the ID field after the Document has been committed, or take necessary precautions if allowing it.

If we want to allow it, then maybe the role of `is_created` needs to be redefined, I don't know. At least, there should be a way to remember the old pk value so that the update can find the document.

I'm not sure this is a critical use case, but if feasible, I'd rather allow it.

Labelling as bug as it leads to an UpdateError exception while nothing tells the user he's doing something wrong.

**Comment by touilleMan** [comment]
I would say modifying the `pk` field (i.e. the `_id` field) should be forbidden given then the `commit` operation would need to create a new document and delete the old one, which is something that can't be done in a single operation (see http://stackoverflow.com/questions/4012855/how-update-the-id-of-one-mongodb-document).

So we should throw an exception if user try such a thing, giving him a tip on what is the "right way" to perform this.

So far, I not really sure about this "right way" ;-)
I'm thinking of providing a `clone` method that would copy everything except the `pk` field. We could also tell the user to use this method when he tries to use copy/deepcopy on the Document object.

What do you think ?

**Comment by lafrech** [comment]
I didn't know about this MongoBD "limitation". Throwing an exception makes sense, then.

As to the `clone()` method, yes, it could be useful.

Let's say that `clone()` **deep**copies everything but ID. Practically, this means that in the general case, a new ID will be created on commit, and if there is some ID field with default/missing attribute (most likely a callable, as a constant would not make sense), then ID is set to its default.

About the `__copy__` / `__deepcopy__`, implementing them is nice, unless they have an unexpected behaviour. Should they copy everything and let the user change the ID manually? Or should they behave like `clone()`?

I'd say copy everything. The ODM provides tools to use objects when dealing with the database. If the ID is an email and I deepcopy the object, I don't want that changed. However, I'm supposed to know that ultimately the object is some sort of database proxy and has restrictions (such as unique ID), so I should be cautious when committing. If I'm not, I'll get an Exception due to unique constraint, just as I would for any other field with a unique constraint. To be consistent, let's not modify ID any more than we change the other unique fields. The `clone()` method allows the user to explicitly specify he want a different doc so he expects something to be done for the ID.

To sum it up:

|   | copy  | deepcopy  | clone  |
|---|---|---|---|
| Document  | copy  | deepcopy  | deepcopy except ID |
| EmbeddedDocument  | copy  | deepcopy | - (AttributeError)|

**Comment by lafrech** [comment]
I sent a PR to fix deepcopy (#157) and another one to add `clone` method (#158).

I didn't need to implement `__copy__`/`__deepcopy__`, I just fixed `__getattr__`.

The `clone` method deepcopies the data proxy's data and replaces `id` with the default id (`missing`, in the general case, unless a default id is provided).
